proposal/prefixed-parent-selector.md

```diff
@@ -77,44 +77,52 @@ Note that this syntax is supported by Less and Stylus, and people have run into
 [Previous Proposal](https://github.com/sass/sass/pull/2723)
 
-## Syntax
-(In terms of [Selectors Level 4 Grammar](https://www.w3.org/TR/selectors-4/#grammar))
-
-```
-SelectorList         ::= ComplexSelectorList#
-ComplexSelectorList  ::= ComplexSelector#
-CompoundSelectorList ::= CompoundSelector#
-SimpleSelectorList   ::= SimpleSelector#
-RelativeSelectorList ::= RelativeSelector#
-ComplexSelector      ::= CompoundSelector [ <combinator>? CompoundSelector ]*
-RelativeSelector     ::= <combinator>? ComplexSelector
-CompoundSelector     ::= ParentSelector? <compound-selector>
-                         | <compound-selector> ParentSelector
-                         | ParentSelector
-SimpleSelector       ::= ParentSelector? <simple-selector>
-                         | <simple-selector> ParentSelector
-                         | ParentSelector
-ParentSelector       ::= '&'
-```
+## Summary
 
-Note that it is not necessary to support the following syntax:
+The goal of this proposal is to add support for `foo&` with certain exceptions that are defined in the ‘syntax’ section. Allowing this placement of the parent selector will provide the ability to write the following syntax:
 ```
-.custom-effect {
-  p&:focus {
-    // styles for <p> when focused
+.foo {
+  div& {
+    ...
   }
+  p& {
+    ...
+  }
+}
+
+// Instead of writing
+
+div.foo {
+ ...
+}
+
+p.foo {
+ ...
 }
 ```
 
-Because the same effect can be achieved with one extra level of nesting:
+## Syntax
+
+Each compound selector may only contain one parent selector `&`.  The parent selector
+may appear by itself, after a `<type-selector>`, or on either side of any `<wq-name>`,
+`<subclass-selector>`, `<pseudo-element-selector>` or `<pseudo-class-selector>`.
+
+(In terms of [Selectors Level 4 Grammar](https://www.w3.org/TR/selectors-4/#grammar))
+
 ```
-.custom-effect {
-  p& {
-    &:focus {
-      // styles for <p> when focused
-    }
-  }
-}
+CompoundSelector            ::= [ ParentSelector
+                                | <type-selector>? <subclass-selector>* PseudoSelectors*
+                                | TypeSelectorWithParent <subclass-selector>* PseudoSelectors*
+                                | <type-selector>? SubclassSelectorsWithParent PseudoSelectors*
+                                | <type-selector>? <subclass-selector>*
+                                  PseudoSelectorsWithParent ]!
+TypeSelectorWithParent      ::= ParentSelector <wq-name> | <type-selector> ParentSelector
+SubclassSelectorsWithParent ::= ParentSelector <subclass-selector>+
+                              | <subclass-selector>+ ParentSelector <subclass-selector>*
+PseudoSelectorsWithParent   ::= ParentSelector PseudoSelectors+
+                              | PseudoSelectors+ ParentSelector PseudoSelectors*
+PseudoSelectors             ::= <pseudo-element-selector> <pseudo-class-selector>*
+ParentSelector              ::= '&'
```

Comment:
I think you're making this more complex than it needs to be. @jquense has the right idea: you want to make `ParentSelector ` a sibling of `TypeSelector`, and you want to make the `ParentSelector` production itself allow an `InterpolatedIdentifier` as a prefix and/or suffix. So something like this:

```
CompoundSelector       ::= InitialSimpleSelector
                         | InitialSimpleSelector? TrailingSimpleSelector+
InitialSimpleSelector  ::= TypeSelector | InterpolatedIdentifier? '&' InterpolatedIdentifier?
TrailingSimpleSelector ::= SubclassSelector | PseudoElementSelector
```

Reply:
So the complexity came from the desire to put the parent selector anywhere inside the compound selector while only allowing there to be one parent selector per compound selector.

Because the definition of `<compound-selector>` allows `<pseudo-class-selector>` to appear after `<pseudo-element-selector>` (the `[ <pseudo-element-selector> <pseudo-class-selector>* ]*` part in https://www.w3.org/TR/selectors-4/#typedef-compound-selector), I assumed the following should be supported:
```
::after {
    ::before& {
        ...
    }
    ::before&:hover {
        ...
    }
    ::before:hover& {
        ...
    }
}
```

Unless I am misunderstanding your suggested syntax (always possible), I don't think it allows for something like `::before&:hover` even though the standard allows for `::before:hover` in a compound selector.

Note that if we allow more than one parent selector per compound selector the syntax would get a lot simpler but it seems weird to support that because it would just allow for duplication in selectors.

Reply:
After playing a little with https://www.sassmeister.com/ I noticed that Sass's idea a `<compound-selector>` is actually much more liberal than https://www.w3.org/TR/selectors-4/#typedef-compound-selector.

For example the following compiles even though the spec says that only `<pseudo-class-selector>` is allowed to follow a `<pseudo-element-selector>`:
```
::before.bar {
    ...
}
::before#foo {
  color: white;
}
```

If the actual Sass definition of `<compound-selector>` is more like the following then I completely understand why you think our proposed syntax is excessively complex:
```
CompoundSelector ::= [ <type-selector>? SubclassSelector>* ]!
SubclassSelector ::= <subclass-selector> | <pseudo-element-selector>
```

Reply:
> Because the definition of `<compound-selector>` allows `<pseudo-class-selector>` to appear after `<pseudo-element-selector>` (the `[ <pseudo-element-selector> <pseudo-class-selector>* ]*` part in https://www.w3.org/TR/selectors-4/#typedef-compound-selector), I assumed the following should be supported:
> 
> ```
> ::after {
>     ::before& {
>         ...
>     }
>     ::before&:hover {
>         ...
>     }
>     ::before:hover& {
>         ...
>     }
> }
> ```

My thought was that this would *not* be allowed, and that a parent selector would (still) only be able to appear at the beginning of a compound selector with an identifier prefix and/or suffix. I can see the argument for allowing it to appear anywhere, but it seems like a much less pressing use-case than combining a parent selector with a type selector. I'd prefer to specify it more narrowly for now and expand it later if we see a compelling use-case.



> After playing a little with https://www.sassmeister.com/ I noticed that Sass's idea a `<compound-selector>` is actually much more liberal than https://www.w3.org/TR/selectors-4/#typedef-compound-selector.
> 
> For example the following compiles even though the spec says that only `<pseudo-class-selector>` is allowed to follow a `<pseudo-element-selector>`:
> 
> ```
> ::before.bar {
>     ...
> }
> ::before#foo {
>   color: white;
> }
> ```
> 
> If the actual Sass definition of `<compound-selector>` is more like the following then I completely understand why you think our proposed syntax is excessively complex:
> 
> ```
> CompoundSelector ::= [ <type-selector>? SubclassSelector* ]!
> SubclassSelector ::= <subclass-selector> | <pseudo-element-selector>
> ```

That's right; this is a result of Sass's general principle to be flexible in the CSS it allows in order to be as forwards-compatible as possible. CSS today forbids peudo-elements anywhere but the end of a selector, but who's to say they won't add a new pseudo-element later that works differently?

Reply:
Our code base actually uses the parent selector following an attribute selector because of the [ElementQuery plugin](https://github.com/tysonmatanich/elementQuery).  This JavaScript plugin uses attribute selectors to generate a media-query like interface for DOM elements which don't extend the full width or height of the viewport.  

The following is a simplified example that hides an inner thumbnail image by default but allows it to appear when the wrapper gets wider.  (Note that the hidden case is the default because ElementQuery runs after DOM load so we want the visible DOM modifications to only appear in the less common, wider case.)
```
.wrapper {
    /* ... */

    .thumbnail {
        display: none;

        [min-width~="17em"]& {
            display: block;
        }
    }

    /* ... */
}
```

We are also using parent selectors after class selectors (e.g. `.my-class&`) and pseudo-selectors (e.g. `:not(.my-class)&`).

I feel like it will be much less surprising to developers if the parent selector is allowed anywhere in a compound selector rather than just before or after an identifier token.  It seems especially weird to me that Sass would support `div&` but not support `:not(div)&`.

Fortunately because of Sass's more liberal interpretation of the location of `<pseudo-element-selector>`the syntax isn't crazy like our first proposal.  Something like this should be sufficient:
```
CompoundSelector ::= [ ParentSelector? <ident-token>? TrailingSelector*
                     | <type-selector>? TrailingSelector* ParentSelector? TrailingSelector* ]!
TrailingSelector ::= <subclass-selector> | <pseudo-element-selector>
ParentSelector   ::= '&'
```

Reply:
I'll bring this use-case up at our design meeting next week and get the team's opinions on it.